### PR TITLE
Storytellers for Dynamic

### PR DESCRIPTION
**Pull request by Putnam3145**
## About The Pull Request

Replaces calm/chaos dynamic voting with storytellers: basically an extra way to weigh antags/events based on various criteria, with storytellers being things that sort using those criteria.

Also implements approval voting and ranked choice for votes, depending on arguments. Storytellers use Schulze scoring of ranked choice ballots by default. Hopefully the uses of these can be expanded over time.

## Why It's Good For The Game

It makes dynamic voting less dichotomous, allowing for more people to get what they want. Implementation of approval voting and ranked choice could also help a bit with stuff; it would be rather easy to swap map voting to approval voting or ranked choice this way, for example.

## Changelog
:cl:
add: Dynamic storytellers, a new voting paradigm for dynamic
add: Support for approval voting and condorcet (ranked choice) voting in server votes
/:cl:

### CODE REVIEW

**Comment by ForrestWick** [comment]
I'm curious what each storyteller will do exactly in terms of what it'll attempt and look for.

**Comment by r4d6** [comment]
You should give them names like in Rimworld.
Exemple of Rimworld's storyteller name

Phoebe Chillax
Cassandra Classic
Randy Random

**Comment by Putnam3145** [comment]
thought about it, went with waaaay more basic at first

**Comment by r4d6** [comment]
Exemples ?

**Comment by Putnam3145** [comment]
image
as of now

**Comment by Trilbyspaceclone** [comment]
Oh hey Eris has these system as well right?

**Comment by Putnam3145** [comment]
Yes, although instead of porting it I made it from scratch based on only a vague idea of how rimworld does it.

File: code/game/gamemodes/dynamic/dynamic_storytellers.dm [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+/datum/dynamic_storyteller/memes
+	name = "Story"
+	desc = "Rules that might lead to fun stories to tell."
+	curve_width = 4
+	property_weights = list("story_potential" = 10)
+
+/datum/dynamic_storyteller/cowabunga
+	name = "Cowabunga"
+	curve_centre = 10
+	desc = "The most chaotic rules only. A validhunter's dream."
```
Comment (Ghommie) [inline code/game/gamemodes/dynamic/dynamic_storytellers.dm]: Saying "A validhunter's dream." is pretty much a sucker to those who complain about antagonists being hunted too much. And considering this also drafts in some potentially dangerous random events, I guess the point is more of a wild ride than a witch hunt.
```suggestion
	desc = "The most chaotic rules only. Expect the unexpectable."
```
P.S. The suggested description is just an example, and a reference to some ugandan meme. 

I'll perhaps review the rest later.

Comment (dzahlus) [reply code/game/gamemodes/dynamic/dynamic_storytellers.dm]: Just change it to "Cowabunga it is"

**Comment by Ghommie** [comment]
Linter checks, from a quick look it doesn't seem voted is a static type list, so you need to state new list variables for voted[C.key] where needed.

File: code/controllers/subsystem/vote.dm [hunk 2]
```diff
@@ -116,17 +169,24 @@ SUBSYSTEM_DEF(vote)
 	log_vote(text)
 	remove_action_buttons()
 	to_chat(world, "\n<font color='purple'>[text]</font>")
+	switch(vote_system)
+		if(APPROVAL_VOTING,PLURALITY_VOTING)
+			for(var/i=1,i<=choices.len,i++)
+				SSblackbox.record_feedback("nested tally","voting",choices[choices[i]],list(vote_title_text,choices[i]))
+		if(RANKED_CHOICE_VOTING)
+			for(var/i=1,i<=voted.len,i++)
+				for(var/j=1,j<=voted[voted[i]].len,j++)
```
Comment (Ghommie) [inline code/controllers/subsystem/vote.dm]: Linter check. define a list L for voted[voted[i]].

File: code/__DEFINES/citadel_defines.dm [hunk 3]
```diff
@@ -119,3 +119,11 @@
...
 //belly sound pref things
 #define NORMIE_HEARCHECK 4
+
+// Voting stuff (move this somewhere else eventually probs)
+#define PLURALITY_VOTING 0
+#define APPROVAL_VOTING 1
+#define RANKED_CHOICE_VOTING 2
+
+// Dynamic stuff (see above, move later)
+#define NO_ASSASSIN (1<<0)
```
Comment (Ghommie) [inline code/__DEFINES/citadel_defines.dm]: I don't like citadel_defines.dm, it's some cringe. :grimacing: 
If you could move these into their appropriate defines files , I'd appreciate it.

File: code/controllers/subsystem/vote.dm [hunk 4]
```diff
@@ -196,27 +243,51 @@ SUBSYSTEM_DEF(vote)
...
-				voted += usr.ckey
-				voted[usr.ckey] = vote
-				choices[choices[vote]]++	//check this
-				return vote
-		else if(vote && 1<=vote && vote<=choices.len)
-			choices[choices[voted[usr.ckey]]]--
-			voted[usr.ckey] = vote
-			choices[choices[vote]]++
-			return vote
+		if(vote && vote >= 1 && vote <= choices.len)
```
Comment (Ghommie) [inline code/controllers/subsystem/vote.dm]:
```suggestion
		if(ISINRANGE(vote, 1, choices.len))
```

File: code/game/gamemodes/dynamic/dynamic_storytellers.dm [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+				drafted_rules[rule] = rule.get_weight() + property_weight
+		else if(mode.threat < rule.cost)
+			SSblackbox.record_feedback("tally","dynamic",1,"Times rulesets rejected due to not enough threat to spend")
+	return drafted_rules
+
+
+/datum/dynamic_storyteller/cowabunga
+	name = "Chaotic"
+	curve_centre = 10
+	desc = "Chaos: high. Variation: high. Likely antags: clock cult, revs, wizard."
```
Comment (Ghommie) [inline code/game/gamemodes/dynamic/dynamic_storytellers.dm:158]: In a future, these sort of descriptions will have to be actually classified on New() based on factors such as their property weights and accessible rulesets for easier maintenance, not gonna lie.

**Review by Ghommie — APPROVED** [review]
I heard there have been a few issues with the testmerge in the past days which should have been resolved by now, so tell me when you think it's ready for merge.

File: code/game/gamemodes/dynamic/dynamic_storytellers.dm (hunk 5 above)
Comment (Putnam3145) [reply code/game/gamemodes/dynamic/dynamic_storytellers.dm:158]: not entirely sure i want them permanently at all

File: code/__DEFINES/citadel_defines.dm (hunk 3 above)
Comment (Putnam3145) [reply code/__DEFINES/citadel_defines.dm]: doing it

**Pull request merged**